sql/planbuilder/scalar.go

```diff
@@ -502,10 +502,64 @@ func (b *Builder) buildBinaryScalar(inScope *scope, be *ast.BinaryExpr) sql.Expr
 	return expr
 }
 
+// typeCoerceComparisonLiteral losslessly expands comparison literals to
+// column types to simplify comparison execution.
+func (b *Builder) typeCoerceComparisonLiteral(left, right sql.Expression) (sql.Expression, sql.Expression) {
+	// TODO if one is get field, other is literal, convert the literal to the get field
+	var leftLit, rightLit *expression.Literal
+	var leftGf, rightGf *expression.GetField
+	switch l := left.(type) {
+	case *expression.GetField:
+		leftGf = l
+	case *expression.Literal:
+		leftLit = l
+	}
+	switch r := right.(type) {
+	case *expression.GetField:
+		rightGf = r
+	case *expression.Literal:
+		rightLit = r
+	}
+
+	var swap bool
+	if leftLit != nil && rightGf != nil {
+		swap = true
+		left, right = right, left
+		rightLit, leftGf = leftLit, rightGf
+	}
+
+	if rightLit != nil && leftGf != nil {
+		if types.IsSigned(left.Type()) && types.IsSigned(right.Type()) {
```

Comment:
Seems kind of over specified? If this is a good idea then it's probably a good idea for other types too

Reply:
I made it more general in a way that I think is easy to understand as safe. Types are similar classes and column type is wider than literal. Crossing those two boundaries is doable but trickier, less common, and can depend on the expression type.